pkg/lib/common.ts

```diff
@@ -64,7 +66,7 @@ export function iterate_data(data, callback, batch = 64 * 1024) {
     }
 }
 
-export function join_data(buffers, binary) {
+export function join_data(buffers: AnyStr[], binary: boolean) {
```

Comment:
No `=> AnyStr` return type? Does tsc have partial specializations, to say that it returns `string` if its input is string, and same for UIint8Array?

Reply:
This is very tricky and is a oft-requested feature in TypeScript.  We can define this function as a generic `<T extends string | Uint8Array>` and say that it takes an array of `T` and returns `T`, but it's not true, because it returns `string` or `Uint8Array` and `T` could be chosen to be a subclass of one of those two things.

This comes up a lot as "exact types" or "type parameter lower bound", and remains an open issue...

Reply:
Also: the places where this is used are places that don't know and don't care about the argument value *or* the return type.  It's a dynamic requirement that this function behaves correctly, of course: we wouldn't want to suddenly start sending Uint8Array on a string channel — but it doesn't actually impact anything that would be statically checked, so it seems like a bit of a waste of time to try so hard to get a more accurate static type.

That being said, I should have a return type annotation here, as discussed.  I'll add that.

Reply:
Done.